demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/SimpleDiversifier.java

```diff
@@ -16,74 +16,98 @@
+                ItemModel te = itemLinked.peek();
+                if (genreInWindow.containsKey(te.getGenre())) {
+                    int toleranceTemp = window - slide;
+                    ItemModel itemStart = new ItemModel();
+                    Iterator<ItemModel> itemModelIterator = itemLinked.iterator();
+                    if (itemModelIterator.hasNext()) {
+                        itemStart = itemModelIterator.next();
+                    }
+                    while (toleranceTemp > 0 && itemModelIterator.hasNext()) {
+                        itemStart = itemModelIterator.next();
+                        toleranceTemp--;
+                    }
+                    int startFound = window - slide;
+
+                    while (startFound < Math.min(tolerance + window - slide, itemLinked.size())
+                            && genreInWindow.containsKey(itemStart.getGenre())
+                            && itemModelIterator.hasNext()) {
+                        startFound++;
+                        itemStart = itemModelIterator.next();
                     }
-                    if (findTarget == Math.min(itemModels.size(), i + window + tolerance)) {
-                        findTarget = i + window;
+                    if (toleranceTemp == itemLinked.size() || toleranceTemp == tolerance + window - slide) {
+                        diverseResult.add(itemLinked.peek());
+                        genreInWindow.put(itemLinked.peek().getGenre(), genreInWindow.get(itemLinked.peek().getGenre()) + 1);
+                        itemLinked.remove();
+                        slide++;
+                        continue;
                     }
-                    itemModels.set(i + m, itemModels.get(findTarget));
-                    itemModels.set(findTarget, item);
-                    int value = keys.containsKey(itemModels.get(i + m).getGenre()) ? keys.get(itemModels.get(i + m).getGenre()) + 1 : 1;
-                    keys.put(itemModels.get(i + m).getGenre(), value);
+                    String targetGenre = itemStart.getGenre();
+                    int value = genreInWindow.containsKey(targetGenre) ? genreInWindow.get(targetGenre) + 1 : 1;
+                    diverseResult.add(itemStart);
+                    genreInWindow.put(targetGenre, value);
+                    itemModelIterator.remove();
                 } else {
-                    keys.put(item.getGenre(), 1);
+                    genreInWindow.put(te.getGenre(), 1);
+                    diverseResult.add(te);
+                    itemLinked.remove();
                 }
+                slide++;
             }
+            stepCount++;
         }
-        return itemModels;
+        return diverseResult;
     }
 
-    public static HashSet<String> groupByType(List<ItemModel> numbers) {
-        HashSet<String> hashset = new HashSet<>();
-        Map<String, List<ItemModel>> map = new HashMap<>();
+    public static Map<String, Integer> groupByType(List<ItemModel> numbers) {
```

Comment:
please do not write this function twice, move this into some utils class or parent class.